skills/contextual-embeddings/guide.ipynb

```diff
@@ -0,0 +1,1372 @@
+      "Total input tokens read from cache: 2825073\n",
+      "Total input token savings from prompt caching: 77.04% of all input tokens used were read from cache.\n",
+      "Tokens read from cache come at a 90 percent discount!\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Load the transformed dataset\n",
+    "with open('data/codebase_chunks.json', 'r') as f:\n",
+    "    transformed_dataset = json.load(f)\n",
+    "\n",
+    "# Initialize the ContextualVectorDB\n",
+    "contextual_db = ContextualVectorDB(\"my_contextual_db\")\n",
+    "\n",
+    "# Load and process the data\n",
+    "#note: consider increasing the number of parallel threads to run this faster, or reducing the number of parallel threads if concerned about hitting your API rate limit\n",
+    "contextual_db.load_data(transformed_dataset, parallel_threads=5)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 360,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 39.53it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@5: 86.37%\n",
+      "Total Score: 0.8637192780337941\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 40.05it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@10: 92.81%\n",
+      "Total Score: 0.9280913978494625\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 39.64it/s]"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@20: 93.78%\n",
+      "Total Score: 0.9378360215053763\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "r5 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 5)\n",
+    "r10 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 10)\n",
+    "r20 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Contextual BM25\n",
+    "\n",
+    "Contextual embeddings is an improvement on traditional semantic search RAG, but we can improve performance further. In this section we'll show you how you can use contextual embeddings and *contextual* BM25 together. While you can see performance gains by pairing these techniques together without the context, adding context to these methods reduces the top-20-chunk retrieval failure rate by 42%.\n",
```

Comment:
Maybe I missed this, but where do you describe the BM25 method? Is there a document you can link to? 

Also, might be worth being a little more verbose about the upgrades you're making here just so it's clear. I.e., We are making 2 upgrades to our approach above:
1. bm25 elaboration
2. RRF elaboration